File: design_patterns_101/Behavioral Patterns/Strategy Pattern/strategyPattern.py
from __future__ import annotations
from abc import ABC, abstractmethod

# Context
class Context:
    def __init__(self, strategy: Strategy) -> None:
        self._strategy = strategy

    def context_interface(self) -> None:
        result = self._strategy.algorithm()
        print(result)

# Strategy interface
class Strategy(ABC):
    @abstractmethod
    def algorithm(self) -> str:
        pass

# Concrete Strategies
class ConcreteStrategyA(Strategy):
    def algorithm(self) -> str:
        return "ConcreteStrategyA algorithm"

class ConcreteStrategyB(Strategy):
    def algorithm(self) -> str:
        return "ConcreteStrategyB algorithm"

# Usage
context = Context(ConcreteStrategyA())
context.context_interface()

context = Context(ConcreteStrategyB())
context.context_interface()

File: design_patterns_101/Behavioral Patterns/Strategy Pattern/strategyPattern.js
// Context
class Context {
    constructor(strategy) {
      this.strategy = strategy;
    }
  
    contextInterface() {
      const result = this.strategy.algorithm();
      console.log(result);
    }
  }
  
  // Strategy
  class Strategy {
    algorithm() {
      throw new Error("Abstract method!");
    }
  }
  
  // Concrete Strategies
  class ConcreteStrategyA extends Strategy {
    algorithm() {
      return "ConcreteStrategyA algorithm";
    }
  }
  
  class ConcreteStrategyB extends Strategy {
    algorithm() {
      return "ConcreteStrategyB algorithm";
    }
  }
  
  // Usage
  const context = new Context(new ConcreteStrategyA());
  context.contextInterface();
  
  const context2 = new Context(new ConcreteStrategyB());
  context2.contextInterface();
  